### DISCUSSION TITLE
Dynamic appends in Eloquent model

### DISCUSSION THREAD

**Discussion by saskaak** [laravel/framework#36782] [Q&A]
What I'd like to do is to only append some attributes when a specific relation has been loaded. I found a solution that relies on undocumented `getArrayableAppends`, which feels a bit hacky to me. As this is primarily for convenience I'm no so hot on introducing hacky, potentially breaking code.

The solution:

```php
class Post extends Model
{
    protected $hidden = ['images'];

    public function images()
    {
        return $this->hasMany(Image::class);
    }

    public function getHeroImageAttribute()
    {
        // Some logic to retrieve the hero image
    }

    public function getOtherImagesAttribute()
    {
        // Some logic to retrieve other images
    }

    // The "magic"
    protected function getArrayableAppends()
    {
        if ($this->relationLoaded('images')) {
            $this->append([
                'hero_image',
                'other_images',
            ]);
        }
        return parent::getArrayableAppends();
    }
}
```

Now I can do this in a controller:

```php
Job::with(['images']);
```

Instead of:

```php
Job::with(['images'])->append(['hero_image', 'other_images'])
```

Is there any other way to achieve this? Am I way off already in even trying to do this? Thoughts?

Edit: To clarify my intent, I specifically want the *array elements* added/removed depending on whether a relationship has been loaded so that it matches how relationships are added.

**Comment by ahinkle** [comment]
Just looking for a bit more clarification 🙂  - is `hero_image` and `other_image` a logic-based relationship on `images`? If so, I think there would be a better way to construct the relationship.  Otherwise, something like this may work:

```php
class Post extends Model
{
    protected $hidden = ['images'];
    public $appends = ['hero_image', 'other_images'];

    public function images()
    {
        return $this->hasMany(Image::class);
    }

    public function getHeroImageAttribute()
    {
        if (! $this->relationLoaded('images')) { 
            return $this->setHidden(array_merge($this->hidden, ['hero_image']));
        }

        // Some logic to retrieve the hero image
    }

    public function getOtherImagesAttribute()
    {
        // Some logic to retrieve other images
    }
}

```

**Reply by saskaak** [reply]
I'm actually using https://github.com/spatie/laravel-medialibrary so I can't change how the relationship is set up.

I just tried your suggestion but it doesn't work. It seems to be too late to hide an attribute at that point. Thanks for the idea though, hadn't thought of that.
